ENH: linearly spaced date_range (GH 20808)

onnoeberhard opened the pull request:
- [X] closes #20808
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

onnoeberhard committed:
ENH: linearly spaced date_range (GH 20808)

onnoeberhard committed:
Merge remote-tracking branch 'upstream/master' into daterange

mroeschke commented on pandas/tests/indexes/datetimes/test_date_range.py:
```diff
@@ -162,6 +162,21 @@ def test_date_range_ambiguous_arguments(self):
         with tm.assert_raises_regex(ValueError, msg):
             date_range(start, end, periods=10, freq='s')
 
+    def test_date_range_convenience_periods(self):
```
Could you also test this with the `tz` arg specified? Would also be good to test a `tz` where there is a day light savings transition between `start` and `end`.

jschendel commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -450,6 +450,7 @@ Other Enhancements
 - Updated ``to_gbq`` and ``read_gbq`` signature and documentation to reflect changes from
   the Pandas-GBQ library version 0.4.0. Adds intersphinx mapping to Pandas-GBQ
   library. (:issue:`20564`)
+- :func:`pandas.core.indexes.datetimes.date_range` now returns a linearly spaced DatetimeIndex if ``start``, ``stop``, and ``periods`` are specified, but ``freq`` is not. (:issue:`20808`)
```
- func:\`date_range\` should suffice here
- DatetimeIndex --> \`\`DatetimeIndex\`\`

jschendel commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2704,7 +2715,36 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
+
+    # See https://github.com/pandas-dev/pandas/issues/20808
+    if freq is None and periods is not None and start is not None \
+            and end is not None:
```
A little bit cleaner to use `if freq is None and com._all_not_none(periods, start, end):`

jschendel commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2704,7 +2715,36 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
+            raise TypeError(msg.format(periods=periods))
+
+        start = Timestamp(start)
+        end = Timestamp(end)
```
`start` and `end` should be normalized if `normalize=True` is passed to `date_range`

jschendel commented:
For consistency, it seems like this behavior should also be implemented for `bdate_range`, `timedelta_range`, and `interval_range`.  Doesn't really make sense for `period_range`, since a freq is always required there, and can probably skip `cdate_range` since it's been deprecated (although not removed).

onnoeberhard committed:
Merge remote-tracking branch 'upstream/master' into daterange

onnoeberhard committed:
ENH: linearly spaced date_range (GH 20808)

onnoeberhard committed:
ENH: linearly spaced date_range (GH 20808)

pep8speaks commented:
Hello @onnoeberhard! Thanks for updating the PR.



 - In the file [`pandas/tests/indexes/datetimes/test_date_range.py`](https://github.com/pandas-dev/pandas/blob/580bca2747b596e2d32e7863b0071be29ba52060/pandas/tests/indexes/datetimes/test_date_range.py), following are the PEP8 issues :

> [Line 189:10](https://github.com/pandas-dev/pandas/blob/580bca2747b596e2d32e7863b0071be29ba52060/pandas/tests/indexes/datetimes/test_date_range.py#L189): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 189:10](https://github.com/pandas-dev/pandas/blob/580bca2747b596e2d32e7863b0071be29ba52060/pandas/tests/indexes/datetimes/test_date_range.py#L189): [E116](https://duckduckgo.com/?q=pep8%20E116) unexpected indentation (comment)
> [Line 193:14](https://github.com/pandas-dev/pandas/blob/580bca2747b596e2d32e7863b0071be29ba52060/pandas/tests/indexes/datetimes/test_date_range.py#L193): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four

onnoeberhard committed:
ENH: linearly spaced date_range (GH 20808)

onnoeberhard commented:
Alright, thank you for all the feedback. I have implemented almost everything mentioned, except for the implementation of a similar feature in the functions `bdate_range`, `timedelta_range` and `interval_range`, because I have not yet used these functions and don't know how to properly implement it in an intuitive way.

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2704,7 +2715,49 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
+           '2018-04-27T00:00:00.000000000'], dtype='datetime64[ns]')
     """
+
+    # See https://github.com/pandas-dev/pandas/issues/20808
```
this needs to go in the DTI constructor itself. we already do some of this validation, it needs to fit in there. Further you don't need to worry about lots of other things that you are repeating, e.g. tz, which are already handled.

jreback reviewed (changes requested):
also would be nice to add this to TDI

onnoeberhard committed:
Merge remote-tracking branch 'upstream/master' into daterange

onnoeberhard committed:
Moved to DTI constructor, removed kwargs functionality

onnoeberhard committed:
Moved to DTI constructor, removed kwargs functionality

onnoeberhard commented:
I have moved everything to the DTI constructor. With that I have removed the functionality to declare kwargs to `to_datetime`, but I think it is much cleaner overall like this.

onnoeberhard committed:
Merge remote-tracking branch 'upstream/master' into daterange

onnoeberhard committed:
Fixed Style Issues

onnoeberhard committed:
Fixed Style Issues

jreback commented:
thank @onnoeberhard
